drivers/adc_ng/include/adc_ng_internal.h

```diff
@@ -0,0 +1,289 @@
+     * @param           handle  Handle of the ADC
+     * @param[in]       chan    The ADC channel to initialize
+     * @param[in]       res     The resolution to select
+     * @param[in        ref     Index of the reference to use (@ref adc_ng_driver_t::refs)
+     *
+     * @retval  0               Success
+     * @retval  -ENXIO          Invalid channel given in @p channel
+     * @retval  -EALREADY       The ADC is already powered and configured
+     * @retval  <0              Other error (see device driver doc)
+     *
+     * @post    When 0 is returned, the channel @p channel is ready take samples
+     * @post    If @p res contains an unsupported resolution, an assert blows up
+     * @post    If `-EALREADY` is returned, the ADC is state remains unchanged
+     * @post    On other error codes the ADC is powered down
+     *
+     * @note    A call to @ref adc_ng_driver_t::off is needed to disable the channel
+     *          again and preserve power
+     */
+    int (*init)(void *handle, uint8_t chan, uint8_t res, uint8_t ref);
+    /**
+     * @brief   Disable the given ADC channel again and bring the ADC into a low
+     *          power state, unless other ADC channels are still onchannel
+     *
+     * @param           handle  Handle of the ADC
+     */
+    void (*off)(void *handle);
+    /**
+     * @brief   Runs a single conversion and returns the sample
+     *
+     * @param           handle  Handle of the ADC
+     * @param[out]      dest    The sample will be stored here
+     *
+     * @return  The result of the conversion
+     *
+     * @pre     The ADC has been initialized, see @ref adc_ng_driver_t::init
+     *
+     * @post    If the ADC is not initialized, an assertion blows up
+     *
+     * @retval  0           Success
+     * @retval  <0          Error (check device driver's doc for error codes)
+     */
+    int (*single)(void *handle, uint32_t *dest);
+#ifdef MODULE_ADC_BURST
+    /**
+     * @brief   Runs a burst conversion acquiring multiple samples in fast
+     *          succession
+     *
+     * @param           handle  Handle of the ADC
+     * @param[out]      dest    Buffer to write the results of the burst read to
+     * @param[in]       num     Number of samples to burst read
+     *
+     * @pre     The channel given in @p channel is currently initialized, see
+     *          @ref adc_ng_driver_t::init
+     *
+     * @post    If the ADC is not initialized, an assertion blows up
+     *
+     * @retval  0           Success
+     * @retval  <0          Error (check device driver's doc for error codes)
+     */
+    int (*burst)(void *handle, uint32_t *dest, size_t num);
+#endif /* MODULE_ADC_BURST */
+    /**
+     * @brief   Bitmap containing the supported ADC resolutions
+     *
+     * If e.g. the resolutions 4bit, 6bit and 8bit are supported it should have
+     * the value `BIT4 | BIT6 | BIT8`. Thus, currently the highest resolution
+     * supported is 31 bit.
```

Comment:
What about 32-bit ADCs? I wonder if it would not be sufficient to specify the highest supported resolution as a value here. All lower resolutions can be derived by a simple right shift.

Reply:
My idea was to use `BIT0` for that once a driver has support for this. But as there are no drivers supporting 32 bit yet in RIOT, I thought I might just get away with only supporting 31 for now. (Note that this is part of the internal stuff and we can freely change this later on without deprecation process.)

But if we choose to use `int32_t` instead of `uint32_t`, it might end up confusing if an ADC supports 32 bit when value can get negative, but only 31 bit when the range starts with 0.

Reply:
> My idea was to use BIT0 for that once a driver has support for this.

Okay, this could be a possible solution. But then it should be documented.

> But if we choose to use int32_t instead of uint32_t ...

I don't see any reason why we should change it to a signed type. The resolution of an ADC is always given in bits independent on how the samples are then encoded. IMHO, having an unsigned bitmap would be fine.

But I still don't understand why we need to define the resolutions that are supported by the hardware component at all. For example, if I use a 16-bit ADC and specify that it supports 16-bit resolution, it might not be used if someone only wants to use 10-bit resolution. This would be an unnecessary restriction. Any higher resolution can be converted to a lower resolution very efficiently.

Reply:
Higher resolutions often have higher conversion times, so if you want a quick series of measurements.
I agree that an unsigned type makes more sense as often one is not interested in the absolute voltage but rather the raw value as it acts as a 'percentage' of the reference voltage.

Reply:
> IMHO, having an unsigned bitmap would be fine.

I would like to avoid this for two reasons:
1. When actually calculating with the samples, it has to be converted to the native integer representation of the MCU anyway. Doing so within the driver avoid code duplication
2. The API would no longer be transparent. When switching to a different ADC of another vendor, the application would need to be modified to handle the internal representation of the samples correctly

> But I still don't understand why we need to define the resolutions that are supported by the hardware component at all.

Often the conversion speed depends on the resolution. If I would like to build e.g. a digital scope I might want to sample as fast as possible, but I would not want to throw away bits. So the lower native resolution might be the best choice here.

I also think that the burst sample mode would greatly profit from using a native resolution. E.g. chances are good that the output of the DMA transfer don't need to be modified later on. If one would have to iterate over all samples to shift them, the performance would be reduced.

> Any higher resolution can be converted to a lower resolution very efficiently.

I can provide a convenience function that selects a supported resolution close to but not smaller than the requested one and returns a shift width. E.g. to be used like this:

```C
uint8_t res = 10, adc = 0, shift;
int16_t ref = 3300; /* 3.3V reference */
if (adc_ng_select_res(adc, &res, &shift)) {
    return -1;
}
if (adc_ng_init(adc, chan, res, &ref)) {
    return -1;
}
int32_t sample;
int retval = adc_ng_sample(adc, &sample));
adc_ng_off(adc);
if (retval) {
    return -1;
}
sample >>= shift;
```

Reply:
> Higher resolutions often have higher conversion time
> Often the conversion speed depends on the resolution. 

It's absolutely clear. But to limit a 16-bit ADC only to this resolution would not be the best solution in all use cases.

> I can provide a convenience function that selects a supported resolution close to but not smaller than the requested one and returns a shift width.

That would solve this problem.

Would it then makes sense that `adc_ng_select_res` optionally returns also the conversion time at the selected resolution would take so that the application could decide whether it wants to use it or to be able to calculate the maximum sample rate?

Reply:
> Would it then makes sense that `adc_ng_select_res` optionally returns also the conversion time at the selected resolution so that the application could decide whether it wants to use it or to be able to calculate the maximum sample rate?

As I use I would totally want this feature. However, this would significantly raise the bar for implementations. But doing so optionally, as you suggested, should be good compromise. Right now, I have no idea how that API should look like. I will think about this.

Reply:
Wouldn't it be a further output parameter?
```c
int adc_ng_select_res(uint8_t adc, uint8_t *res, uint8_t *shift, uint16_t *conv_time_us);
```
The `conv_time` pointer could be `NULL` to indicate that it is not required and `*conv_time_us = 0` could be used to indicate that the driver can't set it.

Reply:
This would only be helpful if the user has settled on the resolution and wants to now the corresponding conversion time. But if the user has a fixed conversion time requirement and wants to now which resolutions are compatible, this would be hard to use.

Reply:
I changed the format so that `BIT0` indicates support for 1 bit resolution and so on until `BIT31` indicating support for 32 bit resolutions. That seemed to be more reasonable than having a special case `BIT0` for 32 bit to me now after having thought about this a bit.

Reply:
:+1: OK, it should be documented very carefully.

Reply:
The good thing is that the tests shows all attributes of the supported drivers, so even if one misses the doc, at least the test will point out the issue :-)

Reply:
I added the helper utility. But the conversion time info is still missing.